### PR TITLE
Discount  applying on Shipping Amount for the rule 'Buy X get Y free (discount amount is Y)'

### LINKED ISSUE #26723: Discount is not applying on Shipping Amount for the rule 'Buy X get Y free (discount amount is Y)'.
### Preconditions (*)
Magento version:2.3.4 
Php version 7.2.14

### Steps to reproduce (*)
1. Create a rule for  'Buy X get Y free (discount amount is Y)
2.  Apply to Shipping Amount set to yes.

### Expected result (*)
Shipping amount should be $10
as discount should apply on shipping amount of 1 qty of item.

### Actual result (*)
Discount is not applying on Shipping Amount of 1 qty of item.

--------------------------------------------------------------------
Screenshot (1)

-----------------------------------------------------------
Screenshot (2)

### PR DESCRIPTION
**Pull request by engcom-Echo**
### Fixed Issues (if relevant)

1. Fixes magento/magento2#26723

### Manual testing scenarios (*)

1. See magento/magento2#26723

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by engcom-Echo** [comment]
@magento run all tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @engcom-Echo,
Could you cover your changes with some types of tests?

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by bradleybrecher** [comment]
@magento give me 2.3.5-p1 instance

**Comment by magento-engcom-team** [comment]
Hi @bradleybrecher. Thank you for your request. I'm working on Magento 2.3.5-p1 instance for you

**Comment by magento-engcom-team** [comment]
Hi @bradleybrecher, here is your [Magento instance](https://i-28839-2-3-5-p1.instances.magento-community.engineering).
Admin access: [https://i-28839-2-3-5-p1.instances.magento-community.engineering/admin_b5c8](https://i-28839-2-3-5-p1.instances.magento-community.engineering/admin_b5c8)
Login: `f3937c7c` Password: `93183be11336`
Instance will be terminated in up to 3 hours.

File: app/code/Magento/SalesRule/Model/Validator.php [hunk 1]
```diff
@@ -481,6 +488,40 @@ private function isValidItemForRule(AbstractItem $item, Rule $rule)
...
     }
 
+    /**
+     * Return discount Qty for all items at Buy_X_Get_Y_Action
+     *
+     * @param Quote $quote
+     * @param Rule $rule
+     * @return float
+     */
+    private function getDiscountQtyAllItemsBuyXGetYAction($quote, $rule)
```
Comment (ihor-sviziev) [inline app/code/Magento/SalesRule/Model/Validator.php]: Can we define argument & return types here?

File: app/code/Magento/SalesRule/Model/Validator.php [hunk 2]
```diff
@@ -481,6 +488,40 @@ private function isValidItemForRule(AbstractItem $item, Rule $rule)
...
+    {
+        $discountAllQty = 0;
+        foreach ($quote->getItems() as $item) {
+            $qty = $item->getQty();
+
+            $x = $rule->getDiscountStep();
```
Comment (ihor-sviziev) [inline app/code/Magento/SalesRule/Model/Validator.php]: could you use more clear names here?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

File: app/code/Magento/SalesRule/Model/Validator.php (hunk 2 above)
Comment (ihor-sviziev) [reply app/code/Magento/SalesRule/Model/Validator.php]: As we discussed - it's buy X get Y, so x and y just named as is and is't ok.

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7746](https://jira.corp.magento.com/browse/ENGCOM-7746) has been created to process this Pull Request

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed
**Manual testing scenario:**
- Create simple product with Price=89$
- Create Cart Price Rules

 
image
image

- Go to storefront
- Add 3 simple products to cart
- Go to Shopping Cart page

**Before:**
:heavy_check_mark: Third product is free
:x: Shipping for third product is remain and added to Order Total. Discount include only product price(89$)
image

**After:**
:heavy_check_mark: Third product is free
:heavy_check_mark: Shipping for third product is free and discount include product price(89$) + shipping price(5$)
image

File: app/code/Magento/SalesRule/Model/Validator.php [hunk 3]
```diff
@@ -385,6 +386,12 @@ public function processShippingAmount(Address $address)
                     }
                     $address->setCartFixedRules($cartRules);
                     break;
+                case Rule::BUY_X_GET_Y_ACTION:
+                    $allQtyDiscount = $this->getDiscountQtyAllItemsBuyXGetYAction($quote, $rule);
```
Comment (engcom-Kilo) [inline app/code/Magento/SalesRule/Model/Validator.php:390]: Seems not need apply discount to Shipping, 
if Shipping method Type is "Per Order" (fixed shipping price for Order).

**Comment by engcom-Kilo** [comment]
@engcom-Echo 
Seems not need apply discount to Shipping,
if Shipping method Handling type is "Per Order" or "Fixed" (fixed shipping price for Order).


Configuration___Settings___Stores___Magento_Admin
Shopping_Cart

File: app/code/Magento/SalesRule/Model/Validator.php (hunk 3 above)
Comment (engcom-Kilo) [reply app/code/Magento/SalesRule/Model/Validator.php:390]: Checked
Your changes are correct.

**Pull request merged**